src/Tokenizer.php

```diff
@@ -897,7 +906,7 @@ private function createNextToken(string $string, string $upper, int $offset, Tok
         // Number (decimal, binary, or hex)
         if (
             preg_match(
-                '/\G(\d+(\.\d+)?|0x[\da-fA-F]+|0b[01]+)($|\s|"\'`|' . $this->regexBoundaries . ')/',
+                $this->nextTokenRegexNumber,
```

Comment:
So to be clear, the performance gain is about avoiding this kind of simple concatenation? Or is there more to it?

Reply:
Let me know if you want to split this PR more. This is one main optimization, but see https://3v4l.org/9jitR/rfc#vgit.master_jit for possible maximal improvement of the improved regex generation from list. In the repro, I created/test worst case input, here in the commit messages, I tested one specific benchmark query only.

Also, the performance benefit of the last commit is huge when the Tokenizer is constructed more than once.

Reply:
I have split this PR now to focus only deduplicated regex generation once per Tokenizer lifetime.

Reply:
@greg0ire can I please have your review here?

Reply:
… ok? There is still a diff here, so I don't understand why you're not answering my question.

Reply:
This PR focus on "Improve tokenizer regex matching" - which has 4 commits related to improving regex matching performance. If you want to furhter splut this PR, please let me know.

Reply:
I did not ask for a split though. I'm asking if the performance gain stems from avoiding that simple concatenation, or if there is more to it. Maybe I don't know, not concanenating helps PHP recognizing that this string never changes, and it compiles the regex only once or something?

Reply:
Each commit message summarizes the performance gain:

![image](https://github.com/doctrine/sql-formatter/assets/2228672/541cbf0c-32a0-41ba-acd5-7b938fc97e09)

> avoiding that simple concatenation

The concatenation cannot be done on compile time, thus it is done on runtime and it needs allocation, per each "next token", which is of course very costly.

Reply:
> Each commit message summarizes the performance gain:

And that's great! I was just wondering how to explain the performance gain. If it's just avoiding the concatenation, then fine.